File: core/metrics.py
"""Analysis math over price history. Pure functions, no I/O, no AI.

Everything here is deterministic given the input series, so the offline
(Streamlit) path produces exactly the same numbers as the online (FastAPI) path.
"""

from __future__ import annotations

import math
from .data import History, Quote


def _pct_returns(closes: list[float]) -> list[float]:
    return [(closes[i] / closes[i - 1] - 1.0)
            for i in range(1, len(closes)) if closes[i - 1]]


def total_return_pct(closes: list[float]) -> float:
    if len(closes) < 2 or not closes[0]:
        return 0.0
    return (closes[-1] / closes[0] - 1.0) * 100


def annualized_volatility_pct(closes: list[float]) -> float:
    """Std-dev of daily returns, scaled to annual (~252 trading days)."""
    rets = _pct_returns(closes)
    if len(rets) < 2:
        return 0.0
    mean = sum(rets) / len(rets)
    var = sum((r - mean) ** 2 for r in rets) / (len(rets) - 1)
    return math.sqrt(var) * math.sqrt(252) * 100


def max_drawdown_pct(closes: list[float]) -> float:
    """Largest peak-to-trough drop, as a negative percent."""
    peak, worst = float("-inf"), 0.0
    for c in closes:
        peak = max(peak, c)
        if peak:
            worst = min(worst, (c / peak - 1.0))
    return worst * 100


def moving_average(closes: list[float], window: int) -> float | None:
    if len(closes) < window:
        return None
    return sum(closes[-window:]) / window


def analyze_history(hist: History) -> dict:
    """Per-symbol stat block from a price series."""
    closes = hist.closes
    ma50 = moving_average(closes, 50)
    ma200 = moving_average(closes, 200)
    last = closes[-1] if closes else None
    return {
        "symbol": hist.symbol,
        "points": len(closes),
        "last": last,
        "total_return_pct": round(total_return_pct(closes), 2),
        "annualized_volatility_pct": round(annualized_volatility_pct(closes), 2),
        "max_drawdown_pct": round(max_drawdown_pct(closes), 2),
        "ma50": round(ma50, 4) if ma50 else None,
        "ma200": round(ma200, 4) if ma200 else None,
        # simple trend signal: price vs the two moving averages
        "trend": _trend_signal(last, ma50, ma200),
    }


def _trend_signal(last, ma50, ma200) -> str:
    if last is None or ma50 is None:
        return "insufficient-data"
    if ma200 is not None:
        if last > ma50 > ma200:
            return "strong-uptrend"
        if last < ma50 < ma200:
            return "strong-downtrend"
    return "uptrend" if last > ma50 else "downtrend"


def performance(dates: list[str], closes: list[float]) -> dict:
    """Trailing returns over standard windows (% change to the latest close)."""
    if len(closes) < 2:
        return {}
    last = closes[-1]
    n = len(closes)

    def ret(idx):
        if idx is None or idx < 0 or idx >= n or not closes[idx]:
            return None
        return round((last / closes[idx] - 1) * 100, 2)

    # ~21 trading days per month.
    out = {
        "1W": ret(n - 1 - 5), "1M": ret(n - 1 - 21), "3M": ret(n - 1 - 63),
        "6M": ret(n - 1 - 126), "1Y": ret(0),
    }
    last_year = dates[-1][:4]
    ytd_idx = next((i for i, d in enumerate(dates) if d[:4] == last_year), None)
    out["YTD"] = ret(ytd_idx)
    return out


def portfolio_summary(quotes: list[Quote], analyses: list[dict]) -> dict:
    """Cross-symbol roll-up — the headline numbers for the dashboard top bar."""
    pes = [q.pe for q in quotes if q.pe]
    vols = [a["annualized_volatility_pct"] for a in analyses if a.get("points", 0) > 1]
    rets = [a["total_return_pct"] for a in analyses if a.get("points", 0) > 1]
    gainers = sum(1 for q in quotes if q.change_pct > 0)
    return {
        "n_symbols": len(quotes),
        "gainers": gainers,
        "losers": len(quotes) - gainers,
        "avg_pe": round(sum(pes) / len(pes), 2) if pes else None,
        "avg_total_return_pct": round(sum(rets) / len(rets), 2) if rets else None,
        "avg_volatility_pct": round(sum(vols) / len(vols), 2) if vols else None,
        "most_volatile": _argextreme(analyses, "annualized_volatility_pct", max),
        "best_performer": _argextreme(analyses, "total_return_pct", max),
        "worst_performer": _argextreme(analyses, "total_return_pct", min),
    }


def _argextreme(analyses: list[dict], key: str, fn):
    candidates = [a for a in analyses if a.get("points", 0) > 1]
    if not candidates:
        return None
    return fn(candidates, key=lambda a: a[key])["symbol"]

File: frontend/src/metrics.js
// Analysis maths over a price series, in the browser.
//
// A port of core/metrics.py — the numbers in the summary bar and quote table.
// Used in direct mode (no backend); src/__tests__/metrics.test.js checks it
// against the Python output so the two never quietly disagree.

function pctReturns(closes) {
  const out = [];
  for (let i = 1; i < closes.length; i++) {
    if (closes[i - 1]) out.push(closes[i] / closes[i - 1] - 1);
  }
  return out;
}

/** How much the price changed from the first close to the last, in %. */
export function totalReturnPct(closes) {
  if (closes.length < 2 || !closes[0]) return 0;
  return (closes[closes.length - 1] / closes[0] - 1) * 100;
}

/**
 * How jumpy the price is: the standard deviation of daily returns, scaled up
 * to a year. There are about 252 trading days a year, and randomness grows
 * with the square root of time — hence the √252.
 */
export function annualizedVolatilityPct(closes) {
  const rets = pctReturns(closes);
  if (rets.length < 2) return 0;
  const mean = rets.reduce((a, b) => a + b, 0) / rets.length;
  const variance = rets.reduce((a, r) => a + (r - mean) ** 2, 0) / (rets.length - 1);
  return Math.sqrt(variance) * Math.sqrt(252) * 100;
}

/** The worst fall from a high point to a later low, as a negative %. */
export function maxDrawdownPct(closes) {
  let peak = -Infinity;
  let worst = 0;
  for (const c of closes) {
    peak = Math.max(peak, c);
    if (peak) worst = Math.min(worst, c / peak - 1);
  }
  return worst * 100;
}

function movingAverage(closes, window) {
  if (closes.length < window) return null;
  let s = 0;
  for (let i = closes.length - window; i < closes.length; i++) s += closes[i];
  return s / window;
}

// Python's round() and JS's Math.round() treat exact .5 ties differently, but
// on real prices ties are vanishingly rare and only move the last digit.
const round = (v, d) => Math.round(v * 10 ** d) / 10 ** d;

function trendSignal(last, ma50, ma200) {
  if (last == null || ma50 == null) return "insufficient-data";
  if (ma200 != null) {
    if (last > ma50 && ma50 > ma200) return "strong-uptrend";
    if (last < ma50 && ma50 < ma200) return "strong-downtrend";
  }
  return last > ma50 ? "uptrend" : "downtrend";
}

/** Per-symbol stat block, same keys as metrics.analyze_history. */
export function analyzeHistory(symbol, closes) {
  const ma50 = movingAverage(closes, 50);
  const ma200 = movingAverage(closes, 200);
  const last = closes.length ? closes[closes.length - 1] : null;
  return {
    symbol,
    points: closes.length,
    last,
    total_return_pct: round(totalReturnPct(closes), 2),
    annualized_volatility_pct: round(annualizedVolatilityPct(closes), 2),
    max_drawdown_pct: round(maxDrawdownPct(closes), 2),
    ma50: ma50 ? round(ma50, 4) : null,
    ma200: ma200 ? round(ma200, 4) : null,
    trend: trendSignal(last, ma50, ma200),
  };
}

function argExtreme(analyses, key, pickMax) {
  const c = analyses.filter((a) => (a.points || 0) > 1);
  if (!c.length) return null;
  // First one wins on a tie, like Python's max()/min().
  return c.reduce((best, a) =>
    (pickMax ? a[key] > best[key] : a[key] < best[key]) ? a : best).symbol;
}

const avg = (xs) => (xs.length ? round(xs.reduce((a, b) => a + b, 0) / xs.length, 2) : null);

/** The headline numbers for the summary bar, same keys as portfolio_summary. */
export function portfolioSummary(quotes, analyses) {
  const ok = analyses.filter((a) => (a.points || 0) > 1);
  const gainers = quotes.filter((q) => q.change_pct > 0).length;
  return {
    n_symbols: quotes.length,
    gainers,
    losers: quotes.length - gainers,
    avg_pe: avg(quotes.map((q) => q.pe).filter(Boolean)),
    avg_total_return_pct: avg(ok.map((a) => a.total_return_pct)),
    avg_volatility_pct: avg(ok.map((a) => a.annualized_volatility_pct)),
    most_volatile: argExtreme(analyses, "annualized_volatility_pct", true),
    best_performer: argExtreme(analyses, "total_return_pct", true),
    worst_performer: argExtreme(analyses, "total_return_pct", false),
  };
}
